rpcserver.go

```diff
@@ -339,19 +345,137 @@ type rpcServer struct {
 
 	wg sync.WaitGroup
 
+	// subServers are a set of sub-RPC servers that use the same gRPC and
+	// listening sockets as the main RPC server, but which maintain their
+	// own independent service. This allows us to expose a set of
+	// micro-service like abstractions to the outside world for users to
+	// consume.
+	subServers []lnrpc.SubServer
+
+	// grpcServer is the main gRPC server that this RPC server, and all the
+	// sub-servers will use to register themselves and accept client
+	// requests from.
+	grpcServer *grpc.Server
+
+	// listenerCleanUp are a set of closures functions that will allow this
+	// main RPC server to clean up all the listening socket created for the
+	// server.
+	listenerCleanUp []func()
+
+	// restServerOpts are a set of gRPC dial options that the REST server
+	// proxy will use to connect to the main gRPC server.
+	restServerOpts []grpc.DialOption
+
+	// tlsCfg is the TLS configu that allows the REST server proxy to
+	// connect to the main gRPC server to proxy all incoming requests.
+	tlsCfg *tls.Config
+
 	quit chan struct{}
 }
 
 // A compile time check to ensure that rpcServer fully implements the
 // LightningServer gRPC service.
 var _ lnrpc.LightningServer = (*rpcServer)(nil)
 
-// newRPCServer creates and returns a new instance of the rpcServer.
-func newRPCServer(s *server) *rpcServer {
-	return &rpcServer{
-		server: s,
-		quit:   make(chan struct{}, 1),
+// newRPCServer creates and returns a new instance of the rpcServer. The
+// rpcServer will handle creating all listening sockets needed by it, and any
+// of the sub-servers that it maintains. The set of serverOpts should be the
+// base level options passed to the grPC server. This typically includes things
+// like requiring TLS, etc.
+func newRPCServer(s *server, macService *macaroons.Service,
+	subServerCgs *subRpcServerConfigs, serverOpts []grpc.ServerOption,
+	restServerOpts []grpc.DialOption,
+	tlsCfg *tls.Config) (*rpcServer, error) {
+
+	var (
+		subServers     []lnrpc.SubServer
+		subServerPerms []lnrpc.MacaroonPerms
+	)
+
+	// Before we create any of the sub-servers, we need to ensure that all
+	// the dependencies they need are properly populated within each sub
+	// server configuration struct.
+	err := subServerCgs.PopulateDependancies(
```

Comment:
Why is this level of indirection needed? Seems like we could just pass these parameters to `subserver.New()` or have a common `subserverConfig` passed to `New()` instead of using reflection.

Reply:
This will still have the problem of needing to pass in more parameters in the future, if new subservers require them in their configs.

Reply:
One possibility is to use a map from `subserver name -> subserverconfig`, see https://github.com/lightningnetwork/lnd/commit/b9c74ff61ca4c3200c552f3259adb7c9bbdd7365

Reply:
That ignores that the sub-servers also may have command line arguments that we may need to parse out which are in the config.

Reply:
See https://github.com/lightningnetwork/lnd/commit/b9c74ff61ca4c3200c552f3259adb7c9bbdd7365#r31173493